sql/src/main/java/io/crate/planner/consumer/SemiJoins.java

```diff
@@ -156,20 +157,24 @@ private static AnalyzedRelation getSource(QueriedRelation rel) {
     }
 
     private static void removeRewriteCandidatesFromWhere(QueriedRelation rel, Collection<Candidate> rewriteCandidates) {
-        rel.querySpec().where().replace(FuncReplacer.mapNodes(f -> {
-            for (Candidate rewriteCandidate : rewriteCandidates) {
-                if (rewriteCandidate.function.equals(f)) {
+        for (Candidate candidate : rewriteCandidates) {
+            java.util.function.Function<? super Symbol, ? extends Symbol> replaceFunction = f -> {
+                if (candidate.function.equals(f)) {
                     return Literal.BOOLEAN_TRUE;
                 }
+                return f;
+            };
+            if (candidate.joinType == JoinType.SEMI) {
+                replaceFunction = FuncReplacer.mapNodes(replaceFunction);
```

Comment:
why is this only required for SEMI joins?

Reply:
Because otherwise the `mapNodes()` calls `processAndMaybeCopy()` which processes the argument of the function (in that case the Not operator)